fdbserver/masterserver.actor.cpp

```diff
@@ -1212,6 +1218,150 @@ ACTOR Future<Void> configurationMonitor( Reference<MasterData> self ) {
+//
+// Returns a map of pair<Epoch, EpochEndVersion> : map<tag, savedVersion>, so that
+// the backup range should be [savedVersion + 1, EpochEndVersion).
+std::map<std::pair<LogEpoch, Version>, std::map<Tag, Version>> getUnfinishedBackup(
+    const UID dbgid, const std::map<LogEpoch, std::map<Tag, Version>>& progress,
+    const std::map<LogEpoch, Version>& epochEndVersions) {
+	std::map<std::pair<LogEpoch, Version>, std::map<Tag, Version>> toRecruit;
+
+	for (const auto& [epoch, value] : progress) {
+		auto versionIt = epochEndVersions.find(epoch);
+		if (versionIt != epochEndVersions.end()) {
+			const Version endVersion = versionIt->second;
+			std::map<Tag, Version> tagVersions;
+			for (const auto& [tag, version] : value) {
+				if (version < endVersion - 1) {
+					tagVersions.insert({ tag, version });
+					TraceEvent("BW", dbgid)
+					    .detail("OldEpoch", epoch)
+					    .detail("Tag", tag.toString())
+					    .detail("Version", version)
+					    .detail("EpochEndVersion", endVersion);
+				}
+			}
+			if (!tagVersions.empty()) {
+				toRecruit[{ epoch, endVersion }] = tagVersions;
+			}
+		} else {
+			TraceEvent("BW", dbgid).detail("SkipEpoch", epoch);
+		}
+	}
+	return toRecruit;
+}
+
+ACTOR static Future<Void> recruitBackupWorkers(Reference<MasterData> self) {
+	if (self->backupWorkers.size() == 0) return Void();
+
+	state LogEpoch epoch = self->cstate.myDBState.recoveryCount;
+	state Future<std::map<LogEpoch, std::map<Tag, Version>>> backupProgress = getBackupProgress(self);
+	state std::map<LogEpoch, Version> epochEndVersions = self->logSystem->getEpochEndVersions();
+
+	state std::vector<Future<BackupInterface>> initializationReplies;
+	const int logRouterTags = self->logSystem->getLogRouterTags();
+	const Version startVersion = self->logSystem->getStartVersion();
+	state int i = 0;
+	for (; i < logRouterTags; i++) {
+		const auto& worker = self->backupWorkers[i % self->backupWorkers.size()];
+		InitializeBackupRequest req(deterministicRandom()->randomUniqueID());
+		req.recruitedEpoch = epoch;
+		req.backupEpoch = epoch;
+		req.routerTag = Tag(tagLocalityLogRouter, i);
+		req.startVersion = startVersion;
+		TraceEvent("BackupRecruitment", self->dbgid)
+		    .detail("WorkerID", worker.id())
+		    .detail("Epoch", epoch)
+		    .detail("BackupEpoch", epoch)
+		    .detail("StartVersion", req.startVersion);
+		initializationReplies.push_back(
+		    transformErrors(throwErrorOr(worker.backup.getReplyUnlessFailedFor(
+		                        req, SERVER_KNOBS->BACKUP_TIMEOUT, SERVER_KNOBS->MASTER_FAILURE_SLOPE_DURING_RECOVERY)),
+		                    master_recovery_failed()));
+	}
+
+	std::map<LogEpoch, std::map<Tag, Version>> progress = wait(backupProgress);
+	std::map<std::pair<LogEpoch, Version>, std::map<Tag, Version>> toRecruit =
+	    getUnfinishedBackup(self->dbgid, progress, epochEndVersions);
+	for (const auto& [epochVersion, tagVersions] : toRecruit) {
```

Comment:
we need to chain recruitment, to avoid popping data still needed by previous generations

Reply:
I added oldestBackupEpoch to LogSystem, which is updated when setting/removing backup workers on the master. The backup worker's popping logic verifies that its backup epoch is no larger than the oldestBackupEpoch.

The restore needs to be careful, because there might be holes in the version range. I.e., restore process needs to verify the backup progress to make sure there are no holes, otherwise it needs to wait.